File: ProyectoMysql/server/DataLayer/ReservaDB.py
from EntityLayer.DespachoEntity import DespachoReservaOPModel
from Utilidades.Entidades.ResponseAPI import ResponseAPI
from Utilidades.Conexion.ErrorData import ErrorData
from Utilidades.Conexion.configMysql import DBProcedure, Restore
from EntityLayer.ReservaEntity import *


class ReservaDB:

            
    def GetMercaderiaMainItems():
        try:
            resulset = DBProcedure().DBProcedureConsult("sp_ReservaMercaderiaListaMain", [])
            list = [ReservaItemModel.Cargar(row) for row in resulset]
            return list
        except Exception as e:
            print(e)
            
    def ReservarMercaderiaDB(Ent: ReservaMercaderiaOPModel):
        try:
   
            Store = "sp_ReservarMercaderiaItem"
            args = []
            args.append(Ent.MercaderiaId)
            args.append(Ent.Cantidad)
            args.append(Ent.OrdenPedidoDetalleId)
            Ent.MercaderiaId = DBProcedure().DBProcedureInsertUpdate(
                Store, args, "v_MercaderiaId"
            )
           
            return Ent
        except Exception as e:
            print(e)
            Restore()

    def ReservaPedido(MercaderiaId : int):
        try:
            args = []
            args.append(MercaderiaId)
            resulset = DBProcedure().DBProcedureConsult("sp_ReservaOrdenPedidoListaMain", args)
            list = [ReservaPedidoModel.Cargar(row) for row in resulset]
            return list
        except Exception as e:
            print(e)

    def ReservaResumen(MercaderiaId : int):
        try:
            args = []
            args.append(MercaderiaId)
            resulset = DBProcedure().DBProcedureConsult("sp_ReservaObtenerResumen", args)
            list = [ReservaResumenModel.Cargar(row) for row in resulset]
            return list
        except Exception as e:
            print(e)

    def ObtenerReservaOPItem(OrdenPedidoId : int):
        try:
            args = []
            args.append(OrdenPedidoId)
            resulset = DBProcedure().DBProcedureConsult("sp_ObtenerReservarOPItem", args)
            list = [DespachoReservaOPModel.Cargar(row) for row in resulset]
            return list
        except Exception as e:
            print(e)

    def ReservarDespachoDB(Ent: ReservaSaveModel):
        try:
   
            Store = "sp_RegistrarReservaDespacho"
            args = []
            args.append(Ent.ReservaId)
            args.append(Ent.OrdenPedidoId)
            args.append(Ent.OrdenPedidoDetalleId)
            args.append(Ent.MercaderiaId)
            args.append(Ent.Cantidad)
            args.append(Ent.StockId)

            Ent.MercaderiaId = DBProcedure().DBProcedureInsertUpdate(
                Store, args, "v_MercaderiaId"
            )
           
            return Ent
        except Exception as e:
            print(e)
            Restore()

File: ProyectoMysql/Api/LogisticStorage/LogisticStorage.DataLayer/ReservaDB.cs
using Framework.Data;
using Framework;
using LogisticStorage.Common;
using LogisticStorage.EntityLayer;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LogisticStorage.DataLayer
{
    public class ReservaDB : BaseDataEntity
    {

        public virtual List<ReservaEntity> GetMercaderiaMainItems()
        {
            try
            {
                StartHelper(false);
                IDataReader dr = (IDataReader)DbDatabase.ExecuteReader(System.Data.CommandType.StoredProcedure, "sp_ReservaMercaderiaListaMain");
                FillSchemeTable(dr);
                List<ReservaEntity> EntityList = new List<ReservaEntity>();

                while (dr.Read())
                {
                    ReservaEntity entity = new ReservaEntity();
                    if (FillFrom<ReservaEntity>(dr, entity)) EntityList.Add(entity);
                    entity.OnLogicalLoaded();
                }

                Helper.Close(dr);
                return EntityList;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public virtual bool ReservarMercaderia(ReservaEntity Ent)
        {
            StartHelper(true);
            try
            {
                ReservarMercaderiaDB(Ent);
            }
            catch (Exception ex)
            {
                Helper.CancelTransaction();
                throw ex;
            }

            Helper.Close();
            return true;
        }

        private bool ReservarMercaderiaDB(ReservaEntity Ent)
        {
            if (Ent.LogicalState == LogicalState.Added || Ent.LogicalState == LogicalState.Updated)
            {
                String storedName = "sp_Reserva_Update";
                if (Ent.LogicalState == LogicalState.Added) storedName = "sp_Reserva_SaveOP";
                DbDatabase.GetStoredProcCommand(storedName);
                DbDatabase.SetTransaction(Helper.DbTransaction);



                DbDatabase.AddParameter(MyUtils.GetOutputDirection(true), "v_MercaderiaId", DbType.Int32, 4, false, 0, 0, Ent.MercaderiaId);
                DbDatabase.AddParameter(MyUtils.GetOutputDirection(false), "v_Cantidad", DbType.Decimal, 4, false, 0, 0, Ent.Cantidad);
                DbDatabase.AddParameter(MyUtils.GetOutputDirection(false), "v_OrdenPedidoDetalleId", DbType.Int32, 4, false, 0, 0, Ent.OrdenPedidoDetalleId);
                int returnValue = DbDatabase.ExecuteNonQuery();
                if (Ent.LogicalState == LogicalState.Added)
                {
                    if (Ent.MercaderiaId <= 0) Ent.MercaderiaId = (Int32)DbDatabase.GetParameterValue("v_ReservaId");
                    Ent.OnLogicalAdded();
                }
                else
                {
                    if (returnValue <= 0) throw new Exception("ErrorDB.UpdateEntity");
                    Ent.OnLogicalUpdate();
                }
            }




            return true;
        }

        public virtual List<ReservaEntity> ReservaPedido(Int32 MercaderiaId)
        {
            try
            {
                StartHelper(false);
                DbDatabase.AddParameter(MyUtils.GetOutputDirection(false), "v_MercaderiaId", DbType.Int32, 4, false, 0, 0, MercaderiaId);
                IDataReader dr = (IDataReader)DbDatabase.ExecuteReader(CommandType.StoredProcedure, "sp_ReservaOrdenPedidoListaMain");
                FillSchemeTable(dr);
                List<ReservaEntity> EntityList = new List<ReservaEntity>();
                while (dr.Read())
                {
                    ReservaEntity entity = new ReservaEntity();
                    if (FillFrom(dr, entity)) EntityList.Add(entity);
                    entity.OnLogicalLoaded();
                }

                Helper.Close(dr);
                return EntityList;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public virtual List<ReservaEntity> ReservaResumen(Int32 MercaderiaId)
        {
            try
            {
                StartHelper(false);
                DbDatabase.AddParameter(MyUtils.GetOutputDirection(false), "v_MercaderiaId", DbType.Int32, 4, false, 0, 0, MercaderiaId);
                IDataReader dr = (IDataReader)DbDatabase.ExecuteReader(CommandType.StoredProcedure, "sp_ReservaObtenerResumen");
                FillSchemeTable(dr);
                List<ReservaEntity> EntityList = new List<ReservaEntity>();
                while (dr.Read())
                {
                    ReservaEntity entity = new ReservaEntity();
                    if (FillFrom(dr, entity)) EntityList.Add(entity);
                    entity.OnLogicalLoaded();
                }

                Helper.Close(dr);
                return EntityList;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public virtual List<ReservaEntity> ObtenerReservaOPItem(Int32 OrdenPedidoId)
        {
            try
            {
                StartHelper(false);
                DbDatabase.AddParameter(MyUtils.GetOutputDirection(false), "v_OrdenPedidoId", DbType.Int32, 4, false, 0, 0, OrdenPedidoId);
                IDataReader dr = (IDataReader)DbDatabase.ExecuteReader(CommandType.StoredProcedure, "sp_ObtenerReservarOPItem");
                FillSchemeTable(dr);
                List<ReservaEntity> EntityList = new List<ReservaEntity>();
                while (dr.Read())
                {
                    ReservaEntity entity = new ReservaEntity();
                    if (FillFrom(dr, entity)) EntityList.Add(entity);
                    entity.OnLogicalLoaded();
                }

                Helper.Close(dr);
                return EntityList;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public virtual bool ReservarDespacho(ReservaEntity Ent)
        {
            StartHelper(true);
            try
            {
                //if (Ent.LogicalState == LogicalState.Deleted) EliminarDB(Ent);
                //else RegistrarDB(Ent);
                ReservarDespachoDB(Ent);
            }
            catch (Exception ex)
            {
                Helper.CancelTransaction();
                throw ex;
            }

            Helper.Close();
            return true;
        }
        private bool ReservarDespachoDB(ReservaEntity Ent)
        {
            //if (Ent.LogicalState == LogicalState.Added || Ent.LogicalState == LogicalState.Updated)
            //{
                Ent.LogicalState = LogicalState.Updated;
                String storedName = "sp_RegistrarReservaDespacho";
                DbDatabase.GetStoredProcCommand(storedName);
                DbDatabase.SetTransaction(Helper.DbTransaction);



                DbDatabase.AddParameter(MyUtils.GetOutputDirection(true), "v_ReservaId", DbType.Int32, 4, false, 0, 0, Ent.ReservaId);
                DbDatabase.AddParameter(MyUtils.GetOutputDirection(false), "v_OrdenPedidoId", DbType.Int32, 4, false, 0, 0, Ent.OrdenPedidoId);
                DbDatabase.AddParameter(MyUtils.GetOutputDirection(false), "v_OrdenPedidoDetalleId", DbType.Int32, 4, false, 0, 0, Ent.OrdenPedidoDetalleId);
                DbDatabase.AddParameter(MyUtils.GetOutputDirection(false), "v_MercaderiaId", DbType.Int32, 4, false, 0, 0, Ent.MercaderiaId);
                DbDatabase.AddParameter(MyUtils.GetOutputDirection(false), "v_Cantidad", DbType.Decimal, 4, false, 0, 0, Ent.Cantidad);
                DbDatabase.AddParameter(MyUtils.GetOutputDirection(false), "v_StockId", DbType.Int32, 4, false, 0, 0, Ent.StockId);


                int returnValue = DbDatabase.ExecuteNonQuery();
                //if (Ent.LogicalState == LogicalState.Added)
                //{
                //    if (Ent.ReservaId <= 0) Ent.ReservaId = (Int32)DbDatabase.GetParameterValue("v_ReservaId");
                //    Ent.OnLogicalAdded();
                //}
                //else
                //{
                //    if (returnValue <= 0) throw new Exception("ErrorDB.UpdateEntity");
                //    Ent.OnLogicalUpdate();
                //}
            //}




            return true;
        }


    }
}
